### PR TITLE
win: rework how events for unmapped windows are handled

### LINKED ISSUE #525: Picom Crashes after some time
### Platform
ArtixLinix

### GPU, drivers, and screen setup
   Rx460
    Mesa 20.2.0-2
   'glxinfo -B'
  ```
 π {picom} next   glxinfo -B                                                                                           ±[next] 
name of display: :0 
display: :0  screen: 0 
direct rendering: Yes 
Extended renderer info (GLX_MESA_query_renderer): 
    Vendor: X.Org (0x1002) 
    Device: AMD Radeon (TM) RX 460 Graphics (POLARIS11, DRM 3.39.0, 5.9.1-7-tkg-bmq, LLVM 10.0.1) (0x67ef) 
    Version: 20.2.0 
    Accelerated: yes 
    Video memory: 4096MB 
    Unified memory: no 
    Preferred profile: core (0x1) 
    Max core profile version: 4.6 
    Max compat profile version: 4.6 
    Max GLES1 profile version: 1.1 
    Max GLES[23] profile version: 3.2 
Memory info (GL_ATI_meminfo): 
    VBO free memory - total: 3576 MB, largest block: 3576 MB 
    VBO free aux. memory - total: 4016 MB, largest block: 4016 MB 
    Texture free memory - total: 3576 MB, largest block: 3576 MB 
    Texture free aux. memory - total: 4016 MB, largest block: 4016 MB 
    Renderbuffer free memory - total: 3576 MB, largest block: 3576 MB 
    Renderbuffer free aux. memory - total: 4016 MB, largest block: 4016 MB 
Memory info (GL_NVX_gpu_memory_info): 
    Dedicated video memory: 4096 MB 
    Total available memory: 8192 MB 
    Currently available dedicated video memory: 3576 MB 
OpenGL vendor string: X.Org 
OpenGL renderer string: AMD Radeon (TM) RX 460 Graphics (POLARIS11, DRM 3.39.0, 5.9.1-7-tkg-bmq, LLVM 10.0.1) 
OpenGL core profile version string: 4.6 (Core Profile) Mesa 20.2.0 
OpenGL core profile shading language version string: 4.60 
OpenGL core profile context flags: (none) 
OpenGL core profile profile mask: core profile 
OpenGL version string: 4.6 (Compatibility Profile) Mesa 20.2.0 
OpenGL shading language version string: 4.60 
OpenGL context flags: (none) 
OpenGL profile mask: compatibility profile 
OpenGL ES profile version string: OpenGL ES 3.2 Mesa 20.2.0 
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.20 

```

### Environment
fork of dwm

### picom version
vgit-9c7f8

```
 π {picom} next   picom --diagnostics                                                                                  ±[next] 
[ 10/25/2020 20:51:01.597 get_cfg WARN ] Dual-kawase blur is not implemented by the legacy backends, you must use the `experimental-backends` option. 
[ 10/25/2020 20:51:01.690 init_render WARN ] Old backends only support blur method "kernel". Your blur setting will not be applied 
**Version:** vgit-9c7f8 
### Extensions: 
* Shape: Yes 
* XRandR: Yes 
* Present: Present 
### Misc: 
* Use Overlay: Yes 
* Config file used: /home/artix7232/.config/picom/picom.conf 
### Drivers (inaccurate): 
AMDGPU, Radeon 
```

### Configuration:
```
 π {~}   cat .config/picom/picom.conf 
backend = "glx"; 
glx-no-stencil = true; 
glx-copy-from-front = false; 
################################# 
# 
# Shadows 
# 
################################# 
# Enabled client-side shadows on windows. 
shadow = true; 
# The blur radius for shadows. (default 12) 
shadow-radius = 8; 
# The left offset for shadows. (default -15) 
shadow-offset-x = -10; 
# The top offset for shadows. (default -15) 
shadow-offset-y = -10; 
# The translucency for shadows. (default .75) 
shadow-opacity = 0.40; 
# Set if you want different colour shadows 
# shadow-red = 0.0; 
# shadow-green = 0.0; 
# shadow-blue = 0.0; 
# The shadow exclude options are helpful if you have shadows enabled. Due to the way picom draws its shadows, certain applications will have visual glitches 
# (most applications are fine, only apps that do weird things with xshapes or argb are affected). 
# This list includes all the affected apps I found in my testing. The "! name~=''" part excludes shadows on any "Unknown" windows, this prevents a visual glitch with the XFWM alt tab switcher. 
shadow-exclude = [ 
        "class_g = 'Cairo-clock'" , 
        "class_g = 'CoverGloobus'", 
        "class_g = 'Tilda'", 
        "name ?= 'gnome-pie'", 
        "class_g = 'Brave'", 
        "class_g = 'Brave-beta'", 
    "class_g = 'brave'", 
        "class_g = 'Dmenu'", 
#       "class_g ?= 'rofi'", 
#       "class_g ?= 'polybar'", 
    "name = 'Notification'", 
    "name = 'Plank'", 
    "name = 'Docky'", 
    "name = 'Kupfer'", 
    "name = 'xfce4-notifyd'", 
    "name *= 'VLC'", 
    "class_g = 'VLC (TopWindow)'", 
    "name *= 'Chromium'", 
    "name *= 'Brave'", 
    "name *= 'Chrome'", 
    "class_g = 'Conky'", 
    "class_g = 'Plank'", 
    "class_g = 'Kupfer'", 
    "class_g = 'Synapse'", 
    "class_g ?= 'Notify-osd'", 
    "class_g ?= 'Cairo-dock'", 
    "class_g ?= 'Xfce4-notifyd'", 
    "class_g ?= 'Xfce4-power-manager'", 
    "_GTK_FRAME_EXTENTS@:c", 
    "_NET_WM_STATE@:32a *= '_NET_WM_STATE_HIDDEN'", 
    "class_g = 'Brave-browser'", 
    "class_g = 'brave-browser'", 
    "class_g = 'KotatogramDesktop'", 
    "class_g = 'code'", 
    "class_g = 'dwmsystray'", 
    "class_g = 'mpv'" 
]; 
# Avoid drawing shadow on all shaped windows (see also: --detect-rounded-corners) 
shadow-ignore-shaped = false; 
# Exclude conditions for background blur. 
################################# 
# 
# Opacity 
# 
################################# 
inactive-opacity = 1; 
active-opacity = 1; 
frame-opacity = 1; 
inactive-opacity-override = false; 
blur-background = true; 
blur-kern = "11x11gaussian"; 
blur-method = "dual_kawase"; 
blur-size = 60; 
################################# 
# 
# Fading 
# 
################################# 
# Fade windows during opacity changes. 
fading = true; 
# The time between steps in a fade in milliseconds. (default 10). 
fade-delta = 1.2; 
# Opacity change between steps while fading in. (default 0.028). 
fade-in-step = 0.02; 
# Opacity change between steps while fading out. (default 0.03). 
fade-out-step = 0.02; 
# Fade windows in/out when opening/closing 
# no-fading-openclose = true; 
# Specify a list of conditions of windows that should not be faded. 
fade-exclude = [ 
      "name *= '- rofi'", 
      "name *= '- Rofi'", 
      "class_g = 'rofi'", 
      "class_g = 'Rofi'", 
      "class_g = 'Brave-browser'", 
      "class_g = 'brave-browser'", 
      "class_g = 'VLC (TopWindow)'", 
      "name *= 'VLC'", 
      "class_g = 'mpv'", 
      "class_g = 'KotatogramDesktop'", 
]; 
################################# 
# 
# Other 
# 
################################# 
# Try to detect WM windows and mark them as active. 
mark-wmwin-focused = true; 
# Mark all non-WM but override-redirect windows active (e.g. menus). 
mark-ovredir-focused = true; 
# Use EWMH _NET_WM_ACTIVE_WINDOW to determine which window is focused instead of using FocusIn/Out events. 
# Usually more reliable but depends on a EWMH-compliant WM. 
use-ewmh-active-win = true; 
# Detect rounded corners and treat them as rectangular when --shadow-ignore-shaped is on. 
detect-rounded-corners = true; 
# Detect _NET_WM_OPACITY on client windows, useful for window managers not passing _NET_WM_OPACITY of client windows to frame windows. 
# This prevents opacity being ignored for some apps. 
# For example without this enabled my xfce4-notifyd is 100% opacity no matter what. 
detect-client-opacity = true; 
# Specify refresh rate of the screen. 
# If not specified or 0, picom will try detecting this with X RandR extension. 
refresh-rate = 0; 
# Vertical synchronization: match the refresh rate of the monitor 
vsync = true; 
# Enable DBE painting mode, intended to use with VSync to (hopefully) eliminate tearing. 
# Reported to have no effect, though. 
dbe = false; 
# Limit picom to repaint at most once every 1 / refresh_rate second to boost performance. 
# This should not be used with --vsync drm/opengl/opengl-oml as they essentially does --sw-opti's job already, 
# unless you wish to specify a lower refresh rate than the actual value. 
#sw-opti = true; 
# Unredirect all windows if a full-screen opaque window is detected, to maximize performance for full-screen windows, like games.
# Known to cause flickering when redirecting/unredirecting windows. 
unredir-if-possible = false; 
# Exclude conditions for background blur. 
blur-background-exclude = [ 
  "window_type = 'dock'", 
  "window_type = 'desktop'", 
  "class_g = 'URxvt'", 
  "class_g = 'slop'", 
  "_GTK_FRAME_EXTENTS@:c", 
  "class_g = 'dmenu'", 
  "class_g = 'brave'", 
]; 
opacity-rule = [ 
    "91:class_g = 'KotatogramDesktop'", 
    "88:class_g = 'Spotify'", 
    "97:class_g = 'Code'", 
    "89:class_g = 'Caprine'" 
]; 
# Specify a list of conditions of windows that should always be considered focused. 
focus-exclude = [ ]; 
# Use WM_TRANSIENT_FOR to group windows, and consider windows in the same group focused at the same time. 
detect-transient = true; 
# Use WM_CLIENT_LEADER to group windows, and consider windows in the same group focused at the same time. 
# WM_TRANSIENT_FOR has higher priority if --detect-transient is enabled, too. 
detect-client-leader = true; 
################################# 
# 
# Window type settings 
# 
################################# 
wintypes: 
{ 
    tooltip = 
    { 
        # fade: Fade the particular type of windows. 
        fade = true; 
        # shadow: Give those windows shadow 
        shadow = true; 
        # opacity: Default opacity for the type of windows. 
        opacity = 0.95; 
        # focus: Whether to always consider windows of this type focused. 
        focus = true; 
    }; 
    utility = 
        { 
                shadow = false; 
        }; 
}; 

```

### Steps of reproduction

1.
2.

### Expected behavior
Not sure.
### Current Behavior
crashes after some time(can't figure out why) and segfaults when running emacs
### Stack trace
```
[ 10/25/2020 20:43:32.653 win_bind_pixmap ERROR ] Failed to get named pixmap for window 0x03800006(emacs)
[1]    31314 segmentation fault  picom --config ~/.config/picom/picom.conf --experimental-backends
```

### PR DESCRIPTION
**Pull request by yshui**
Make unmapped window events work mostly like a mapped window, except
flags set on unmapped windows aren't processed until the window is
mapped.

Hopefully this unifies some of the code paths and reduce corner cases.

Should fix #525

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

## TODO
- [x] Add a test case for #525

### CODE REVIEW

**Comment by yshui** [comment]
@tryone144 would appreciate if you can give this a look, thanks.

File: src/win.c [hunk 1]
```diff
@@ -2226,23 +2277,6 @@ void map_win_start(session_t *ps, struct managed_win *w) {
...
-	xcb_get_geometry_cookie_t gcookie = xcb_get_geometry(ps->c, w->base.id);
-	xcb_get_geometry_reply_t *g = xcb_get_geometry_reply(ps->c, gcookie, NULL);
-
-	if (!g) {
-		log_error("Failed to get the geometry of window %#010x", w->base.id);
-		return;
-	}
-
-	w->g = *g;
-	free(g);
```
Comment (yshui) [inline src/win.c:2242]: geometry is now updated in `configure_win`

File: src/win.c [hunk 2]
```diff
@@ -2226,23 +2277,6 @@ void map_win_start(session_t *ps, struct managed_win *w) {
...
-
-	win_on_win_size_change(ps, w);
```
Comment (yshui) [inline src/win.c:2244]: this is now called in `win_process_update_flags`

File: src/win.c [hunk 3]
```diff
@@ -2254,14 +2288,6 @@ void map_win_start(session_t *ps, struct managed_win *w) {
 	// XXX Can we assume map_state is always viewable?
 	w->a.map_state = XCB_MAP_STATE_VIEWABLE;
 
-	win_update_screen(ps->xinerama_nscrs, ps->xinerama_scr_regs, w);
```
Comment (yshui) [inline src/win.c:2257]: this is now updated in `configure_win`

File: src/win.c [hunk 4]
```diff
@@ -2270,32 +2296,13 @@ void map_win_start(session_t *ps, struct managed_win *w) {
...
 
-	// Detect client window here instead of in add_win() as the client
-	// window should have been prepared at this point
-	if (!w->client_win) {
-		win_recheck_client(ps, w);
-	} else {
-		// Re-mark client window here
-		win_mark_client(ps, w, w->client_win);
-	}
-	assert(w->client_win);
```
Comment (yshui) [inline src/win.c:2281]: this is now updated in `win_process_update_flags`

File: src/win.c [hunk 5]
```diff
@@ -2270,32 +2296,13 @@ void map_win_start(session_t *ps, struct managed_win *w) {
...
 	win_update_focused(ps, w);
 
-	// Update opacity and dim state
-	win_update_opacity_prop(ps, w);
-
-	// Check for _COMPTON_SHADOW
-	win_update_prop_shadow_raw(ps, w);
-
-	// Many things above could affect shadow
-	win_determine_shadow(ps, w);
```
Comment (yshui) [inline src/win.c:2297]: ditto

File: src/win.c [hunk 6]
```diff
@@ -2312,16 +2319,10 @@ void map_win_start(session_t *ps, struct managed_win *w) {
...
 
-	// We stopped listening on ShapeNotify events
-	// when the window is unmapped (XXX we shouldn't),
-	// so the shape of the window might have changed,
-	// update. (Issue #35)
-	//
-	// Also this sets the WIN_FLAGS_IMAGES_STALE flag so later in the critical section
+	// Sets the WIN_FLAGS_IMAGES_STALE flag so later in the critical section
 	// the window's image will be bound
-	win_update_bounding_shape(ps, w);
```
Comment (yshui) [inline src/win.c:2322]: ditto

**Comment by yshui** [comment]
Observed a crash on this. investigating...

File: src/win.c [hunk 7]
```diff
@@ -2254,48 +2290,16 @@ void map_win_start(session_t *ps, struct managed_win *w) {
...
-		win_mark_client(ps, w, w->client_win);
-	}
-	assert(w->client_win);
-
 	log_debug("Window (%#010x) has type %s", w->base.id, WINTYPES[w->window_type]);
 
 	// TODO(yshui) can we just replace calls below with win_on_factor_change?
 
 	// Update window focus state
 	win_update_focused(ps, w);
```
Comment (tryone144) [inline src/win.c:2301]: Do we have to run all the logic for `win_on_factor_change` when we map the window? Or can we rely on `recheck_focus()` being called in the critical section to take care of that?

File: src/win.c [hunk 8]
```diff
@@ -804,15 +817,9 @@ static void win_set_shadow(session_t *ps, struct managed_win *w, bool shadow_new
...
-		// Unmapped window shouldn't have any images
-		w->shadow = shadow_new;
-		assert(!w->shadow_image);
-		assert(!w->win_image);
-		assert(win_check_flags_all(w, WIN_FLAGS_IMAGES_NONE));
-		return;
-	}
+	// We don't handle property updates of non-visible windows until they are mapped.
+	assert(w->state != WSTATE_UNMAPPED && w->state != WSTATE_DESTROYING &&
+	       w->state != WSTATE_UNMAPPING);
```
Comment (tryone144) [inline src/win.c:822]: We should add similar checks in the other functions that should only be called in the critical section (i.e. "mapped" windows). Quick search yielded `win_on_win_size_change()` and `win_update_bounding_shape()`.

**Review by tryone144 — APPROVED** [review]
Added some comments to clarify. But looks good overall.
- Some questions left regarding `destroy_backend()`.
- We could add assertion checks to, e.g., `win_on_win_size_change()` to be only called on "mapped" windows like already done in `win_set_shadow()`.
- Is the xinerama screen properly initialized in `fill_win` or by a configure event, or does this introduce a regression?
- Todo note in `map_win_start()` regarding `win_on_factor_change()`.

File: src/win.c (hunk 1 above)
Comment (yshui) [reply src/win.c:2242]: Yep, you can find the relevant changes in this commit.

File: src/win.c (hunk 3 above)
Comment (yshui) [reply src/win.c:2257]: Good point, let me check.

Comment (yshui) [reply src/win.c:2257]: I did forget about this one. Thanks.

File: src/win.c (hunk 7 above)
Comment (yshui) [reply src/win.c]: I don't want to change that in this PR. We can break it up later if we want to.

File: src/picom.c [hunk 9]
```diff
@@ -405,8 +405,12 @@ static void destroy_backend(session_t *ps) {
 			if (w->state == WSTATE_MAPPED) {
 				win_release_images(ps->backend_data, w);
 			} else {
+				// Unmapped windows could still have shadow images.
+				// And they might have image stale flags set by events.
 				assert(!w->win_image);
-				assert(!w->shadow_image);
+				win_clear_flags(
+				    w, WIN_FLAGS_PIXMAP_STALE | WIN_FLAGS_SHADOW_STALE);
+				win_release_images(ps->backend_data, w);
```
[the remark this answers is not in the document]
Comment (yshui) [reply src/picom.c:413]: Interesting, I wonder why. I think `destroy_backend` can only be reached after the critical section, so the stale flags should already be handled for mapped windows.

File: src/win.c [hunk 10]
```diff
@@ -2254,48 +2292,11 @@ void map_win_start(session_t *ps, struct managed_win *w) {
...
-		win_mark_client(ps, w, w->client_win);
-	}
-	assert(w->client_win);
-
 	log_debug("Window (%#010x) has type %s", w->base.id, WINTYPES[w->window_type]);
 
-	// TODO(yshui) can we just replace calls below with win_on_factor_change?
-
-	// Update window focus state
-	win_update_focused(ps, w);
```
Comment (yshui) [inline src/win.c:2288]: this is called in `win_process_update_flags` -> `win_on_factor_change`

File: src/win.c [hunk 11]
```diff
@@ -2305,23 +2306,15 @@ void map_win_start(session_t *ps, struct managed_win *w) {
 	log_debug("Window %#010x has opacity %f, opacity target is %f", w->base.id,
 	          w->opacity, w->opacity_target);
 
-	win_determine_blur_background(ps, w);
```
Comment (yshui) [inline src/win.c:2308]: this is called in `win_process_update_flags` -> `win_on_factor_change`

File: src/win.c [hunk 12]
```diff
@@ -2254,48 +2290,16 @@ void map_win_start(session_t *ps, struct managed_win *w) {
...
-	// changes are lost
-	xcb_change_window_attributes(
-	    ps->c, w->base.id, XCB_CW_EVENT_MASK,
-	    (const uint32_t[]){determine_evmask(ps, w->base.id, WIN_EVMODE_FRAME)});
-
-	// Get notification when the shape of a window changes
-	if (ps->shape_exists) {
-		xcb_shape_select_input(ps->c, w->base.id, 1);
-	}
-
```
[the remark this answers is not in the document]
Comment (yshui) [reply src/win.c:2269]: That's correct.

**Comment by yshui** [comment]
I deleted two more update calls from `map_win_start`

File: src/win.c (hunk 11 above)
Comment (tryone144) [reply src/win.c:2308]: Right, `WIN_FLAGS_FACTOR_CHANGED` is set in `fill_win` and has been triggered by the corresponding event else.

File: src/win.c (hunk 8 above)
Comment (yshui) [reply src/win.c:822]: Added, thanks.

**Comment by yshui** [comment]
@tryone144 Actually, my impression about what caused #525 is wrong. It's not a window received an update when it's unmapped, it happens because a window's pixmap is invalidated, then the window immediately got unmapped.

This PR should still fix the problem though.

File: src/picom.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (tryone144) [reply src/picom.c:413]: - `paint_preprocess()` -> `unredirect()` -> `destroy_backend()` is called after the critical section before rendering and should be safe.
- `session_destroy()` -> `unredirect()` -> `destroy_backend()` is called in `main` after `session_run()` returns. Are we safe there?
- `handle_root_flags()` -> `configure_root()` -> `destroy_backend()` is called IN the critical section BEFORE the stale flags have been handled. This might fail if pixmaps get invalidated during a root-change (i.e. screen size changes?). I have a testcase for that, but I can't get the timing right to trigger this.

**Comment by tryone144** [comment]
@yshui I think the failure in `win_on_win_size_change` in #525 is just a secondary issue. The main problem was the handling of events for windows that just became unmapped.
We just never got to the point where we refreshed the invalidated pixmaps since win_size is handled before and failed.

**Comment by yshui** [comment]
@tryone144 

> The main problem was the handling of events for windows that just became unmapped.

Right. It's just my impression of what happens first was wrong.

File: src/picom.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (yshui) [reply src/picom.c:413]: You are right, 2 and 3 are not safe.

I don't see a good solution beside always clear the stale flags, so that's what I'll do.

[the remark this answers is not in the document]
Comment (yshui) [reply src/picom.c:413]: Can you post your testcase for 3 somewhere? Thanks.

[the remark this answers is not in the document]
Comment (tryone144) [reply src/picom.c:413]: > Can you post your testcase for 3 somewhere? Thanks.

https://gist.github.com/tryone144/1103f5107448fecf5e7e02a6cf4ffb02

[the remark this answers is not in the document]
Comment (yshui) [reply src/picom.c]: @tryone144

File: src/picom.c [hunk 13]
```diff
@@ -402,12 +402,17 @@ static void destroy_backend(session_t *ps) {
...
-				assert(!w->win_image);
-				assert(!w->shadow_image);
-			}
+			// Unmapped windows could still have shadow images, but not pixmap
+			// images
+			assert(!w->win_image || w->state != WSTATE_UNMAPPED);
+			// Windows can still have stale flags set. For mapped windows,
+			// this can happen when destroy_backend is called before the stale
+			// flags are handled (e.g. when destroy_backend is called in
+			// sessionsession_destroy, or configure_root); for unmapped
```
Comment (tryone144) [inline src/picom.c:411]: Typo `sessionsession_destroy`

File: src/picom.c [hunk 14]
```diff
@@ -402,12 +402,17 @@ static void destroy_backend(session_t *ps) {
...
+			// windows, this is because their stale flags aren't handled until
+			// they are mapped.
+			win_clear_flags(w, WIN_FLAGS_PIXMAP_STALE | WIN_FLAGS_SHADOW_STALE);
```
Comment (tryone144) [inline src/picom.c:414]: We have the `WIN_FLAGS_IMAGES_STALE` version for `PIXMAP | SHADOW`.

File: src/picom.c [hunk 15]
```diff
@@ -495,25 +500,12 @@ static bool initialize_backend(session_t *ps) {
...
-				}
-
-				w->flags |= WIN_FLAGS_PIXMAP_STALE;
-				ps->pending_updates = true;
-			}
+			// We need to reacquire image
+			log_debug("Marking window %#010x (%s) for update after "
+			          "redirection",
+			          w->base.id, w->name);
+			win_set_flags(w, WIN_FLAGS_PIXMAP_STALE | WIN_FLAGS_SHADOW_STALE);
```
Comment (tryone144) [inline src/picom.c:507]: We have the `WIN_FLAGS_IMAGES_STALE` version for `PIXMAP | SHADOW`.

**Review by tryone144 — APPROVED** [review]
Just nitpicking. :wink:

File: src/picom.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (yshui) [reply src/picom.c]: Hmm, looking at the code, I think IMAGES_STALE flags are only set by `win_process_update_flags`, and immediately handled later in `win_process_image_flags`; except for one case, where `map_win_start` is called in `handle_new_windows`.

That seems to be the only case where it's possible for `destroy_backend` to see a image stale flag.

[the remark this answers is not in the document]
Comment (yshui) [reply src/picom.c]: Looks like with this PR, we don't have to call `map_win_start` in `handle_new_windows` anymore. So we can eliminate case 3 (if my previous point is correct).

[the remark this answers is not in the document]
Comment (yshui) [reply src/picom.c]: @tryone144 i'm going to merge this for now.
